crates/next-error-code-swc-plugin/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+                                }))),
+                                PropOrSpread::Prop(Box::new(Prop::KeyValue(KeyValueProp {
+                                    key: PropName::Ident("enumerable".into()),
+                                    value: Box::new(Expr::Lit(Lit::Bool(Bool {
+                                        span: new_error_expr.span,
+                                        value: false,
+                                    }))),
+                                }))),
+                            ],
+                        })), // { value: "$code", enumerable: false }
+                    },
+                ],
+                type_args: None,
+                ctxt: new_error_expr.ctxt,
+            });
+        }
+    }
+}
+
+#[plugin_transform]
+pub fn process_transform(
+    mut program: Program,
+    _metadata: TransformPluginProgramMetadata,
+) -> Program {
+    let errors_json = fs::read_to_string("/cwd/errors.json")
+        .unwrap_or_else(|e| panic!("failed to read errors.json: {}", e));
+    let errors: HashMap<String, String> = serde_json::from_str(&errors_json)
+        .unwrap_or_else(|e| panic!("failed to parse errors.json: {}", e));
+
+    let mut visitor = TransformVisitor { errors };
+
+    visitor.visit_mut_program(&mut program);
+    program
+}
+
+test_inline!(
+    Default::default(),
+    |_| visit_mut_pass(TransformVisitor {
+        errors: HashMap::from([
+            ("1".to_string(), "Failed to fetch user %s: %s".to_string()),
+            ("2".to_string(), "Request failed: %s".to_string()),
+            ("3".to_string(), "Generic error".to_string()),
+            ("4".to_string(), "Empty error".to_string()),
+            (
+                "5".to_string(),
+                "Pattern should define hostname but found\n%s".to_string()
+            ),
+        ]),
+    }),
+    realistic_api_handler,
+    // Input codes
+    r#"
+async function fetchUserData(userId) {
+    try {
+        const response = await fetch(`/api/users/${userId}`);
+        if (!response.ok) {
+            throw new Error(`Failed to fetch user ${userId}: ${response.statusText}`);
+        }
+        return await response.json();
+    } catch (err) {
+        throw new Error(`Request failed: ${err.message}`);
+    }
+}
+
+function test1() {
+    throw Error("Generic error");
+}
+
+function test2() {
+    throw Error();
+}
+
+function test3() {
+    throw new Error("Generic error");
+}
+
+function test4() {
+    throw new Error();
+    throw new Error("Pattern should define hostname but found\n" + JSON.stringify(pattern));
+}"#,
+    // Output codes after transformed with plugin
+    r#"
+async function fetchUserData(userId) {
+    try {
+        const response = await fetch(`/api/users/${userId}`);
+        if (!response.ok) {
+            throw Object.defineProperty(new Error(`Failed to fetch user ${userId}: ${response.statusText}`), "__NEXT_ERROR_CODE", {
+                value: "E1",
+                enumerable: false
```

Comment:
We could consider making it enumerable with a more human-readable property name like `code` and then go with `Next1` for error codes so that they're distinguished from other   error codes. That way the error code would automatically show up when logged to the console

Reply:
If we have a use case for that, we can consider it. Right now, because we only use the code for telemetry, keeping it implicit should be fine.

Reply:
If we have a use case for that, we can consider it. Right now, because we only use the code for telemetry, keeping it implicit should be fine.